streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -409,32 +407,28 @@ public void resume() {
         return committableOffsets;
     }
 
+    /**
+     * This should only be called if the attempted commit succeeded for this task
+     */
     @Override
     public void postCommit() {
         commitRequested = false;
         commitNeeded = false;
 
         switch (state()) {
             case RESTORING:
-                writeCheckpointIfNeed();
+            case SUSPENDED:
+                maybeWriteCheckpoint();
 
                 break;
 
             case RUNNING:
-                if (!eosEnabled) { // if RUNNING, checkpoint only for non-eos
-                    writeCheckpointIfNeed();
+                if (!eosEnabled) {
+                    maybeWriteCheckpoint();
                 }
 
                 break;
 
-            case SUSPENDED:
-                writeCheckpointIfNeed();
-                // we cannot `clear()` the `PartitionGroup` in `suspend()` already, but only after committing,
-                // because otherwise we loose the partition-time information
-                partitionGroup.clear();
```

Comment:
I think we still need to make this call -- in eager rebalancing, we suspend a task when we get a partition revoked. For this case, we "forget" the current offset within the consumer and thus need to clear the partition grouper. Otherwise, we might read the data a second time, if the partition is reassigned (what would violate EOS).

Reply:
I see. So we should only clear it here, and not in `close` 

Just curious, why do we "forget" the current offset? I mean, haven't we just committed the current offset before suspending (and if that failed we would close all tasks right away). Maybe I'm misunderstanding what you mean

Reply:
The consumer tracks offset internal, however, we buffer data in our internal queue. Thus, the offset tracked by the consumer, might be larger than the offset we commit (we take the offset we commit not from the consumer, but it's based on the records we did take out of the queue and processed).

In eager rebalancing, the consumer clears its internal state if a partition in revoked (and we only suspend the task), including the tracked offsets. If the partition in re-assigned, the consumer fetches the last committed offset to start fetching. Thus, if we don't clear the queue, we might fetch same data that is already in the queue a second time.

Does this make sense?

Reply:
Ah that's a good catch. Makes sense to me.

Reply:
Got it, thanks for the explanation. I'll move it back to `postCommit` with a note

Reply:
I'm not 100% certain that the Consumer _does_ clear its internal buffer on revocation. At least, I couldn't find it in the code, but maybe I'm looking in the wrong place. 

Not arguing we shouldn't clear the partition group here, was just wondering about this for my own sake. Hm

Reply:
Not 100% familiar with the consumer code, but in `SubscriptionState#assignFromSubscribed` new `TopicPartitionState` are created with `position = null`.

Reply:
Ah that seems right. Thanks!